tap/api/go.mod

```diff
@@ -0,0 +1,3 @@
+module github.com/up9inc/mizu/tap/api
```

Comment:
@nimrod-up9 have you ever tried to remove this `go.mod` file and run `make agent`?

Let's say you have this diff:

```diff
diff --git a/tap/api/go.mod b/tap/api/go.mod
deleted file mode 100644
index d5379a1..0000000
--- a/tap/api/go.mod
+++ /dev/null
@@ -1,3 +0,0 @@
-module github.com/up9inc/mizu/tap/api
-
-go 1.16
```

You would get:

```bash
$ make agent
building mizu agent ..
go: github.com/up9inc/mizu/tap/api@v0.0.0 (replaced by ../tap/api): reading ../tap/api/go.mod: open /home/mertyildiran/Documents/up9/mizu/tap/api/go.mod: no such file or directory
make: *** [Makefile:36: agent] Error 1
```

or for this diff:

```diff
diff --git a/agent/go.mod b/agent/go.mod
index 15884fa..f1c89f7 100644
--- a/agent/go.mod
+++ b/agent/go.mod
@@ -30,5 +30,3 @@ require (
 replace github.com/up9inc/mizu/shared v0.0.0 => ../shared
 
 replace github.com/up9inc/mizu/tap v0.0.0 => ../tap
-
-replace github.com/up9inc/mizu/tap/api v0.0.0 => ../tap/api
diff --git a/tap/api/go.mod b/tap/api/go.mod
deleted file mode 100644
index d5379a1..0000000
--- a/tap/api/go.mod
+++ /dev/null
@@ -1,3 +0,0 @@
-module github.com/up9inc/mizu/tap/api
-
-go 1.16
```

You would get:

```bash
$ make agent
building mizu agent ..
go: github.com/up9inc/mizu/tap/api@v0.0.0: missing go.sum entry; to add it:
	go mod download github.com/up9inc/mizu/tap/api
make: *** [Makefile:36: agent] Error 1
```

and running `go mod download github.com/up9inc/mizu/tap/api` does not resolve it.

`github.com/up9inc/mizu/tap/api` is a shared Go module that's used by four different compilation units:

- Mizu Agent (driver program)
- HTTP extension (Go plugin)
- AMQP extension (Go plugin)
- Kafka extension (Go plugin)

There is no `go.sum` file since there are no dependencies for the `github.com/up9inc/mizu/tap/api` module.

_Please tell me if I'm missing something._

Reply:
1. About missing `go.sum`, I understand your answer. Thanks.
2. Here there is no need to copy all of `tap/api`. Only copy the `go.mod`. This is done in order to cache `go mod download` in the docker build. If we copy all of `tap/api`, then every change in `tap/api/*` invalidates the cache. If we only copy `tap/api/go.mod`, then only changes in `tap/api/go.mod` invalidate the cache. I pushed a fix.
3. I still don't understand why `tap/api` is a module instead of a package. Is it in order for extensions to be able to depend on `tap/api` without depending on `tap`?

Reply:
@nimrod-up9 I guess 2 is resolved with https://github.com/up9inc/mizu/pull/224/commits/ccf5dc2c16b5ea4f95ef4d45249ba13ef2e3282c ?

For 3: Yes, the goal is not depending on `tap` package from the extensions but instead depending on a small, well defined `tap/api` package. If the extensions depend on `tap` package it would introduce problems and also the `tap` would be recompiled into each shared object. Therefore `tap/api` is defined as a module.

Reply:
Thanks, got it. That's a good design.
I was a little confused by the idea of nested modules (both tap and tap/api are modules), but I guess I will get used to it. I don't see any issue with it.
Resolved.